pkg/kubernetes/tls/certs.go

```diff
@@ -0,0 +1,100 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package tls
+
+import (
+	"context"
+	"github.com/devfile/devworkspace-operator/internal/cluster"
+	"github.com/devfile/devworkspace-operator/webhook/server"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+type GenCertParams struct {
+	RequesterName string
+	Domain        string
+	Namespace     string
+	TLSSecretName string
+	CASecretName  string
+}
+
+// TLS related constants
+const (
+	TLSSelfSignedCertificateSecretName = "devworkspace-self-signed-certificate"
+)
+
+func GenerateCerts(client crclient.Client, ctx context.Context, params GenCertParams) error {
+	// Remove CA certificate secret if any
+	err := removeCACertificate(client, params.CASecretName, params.Namespace)
+	if err != nil {
+		return err
+	}
+
+	jobName := params.RequesterName + "-gen-cert"
+
+	jobEnvVars := map[string]string{
+		"DOMAIN":                         params.Domain,
+		"CHE_NAMESPACE":                  params.Namespace,
+		"CHE_SERVER_TLS_SECRET_NAME":     params.TLSSecretName,
+		"CHE_CA_CERTIFICATE_SECRET_NAME": params.CASecretName,
+	}
+
+	job, err := getSpecJob(server.WebhookServerSAName, jobName, params.Namespace, jobEnvVars)
+	if err != nil {
+		return err
+	}
+
+	err = cluster.SyncJobToCluster(client, ctx, job)
+	if err != nil {
+		return err
+	}
+
+	// Wait a maximum of 60 seconds for the job to be completed
+	err = cluster.WaitForJobCompletion(client, jobName, params.Namespace, 60*time.Second)
```

Comment:
I wonder if 60 seconds is long enough, given how slow we've seen some image pulls to be

Reply:
I've bumped it up to 300 seconds. If theres a way in the kubernetes go code to check if an image is being pulled I could technically do
```
waitInfinite until image is pulled
waitForJobCompletion
```

Reply:
i think at 5 minutes we're in nightmare edge-case territory :)